scanpipe/pipes/d2d.py

```diff
@@ -524,8 +590,55 @@ def match_purldb(project, extensions, matcher_func, logger=None):
     last_percent = 0
     start_time = timer()
     matched_count = 0
+    resources_by_sha1 = defaultdict(list)
+    package_data_by_purldb_urls = {}
 
     for resource_index, to_resource in enumerate(resource_iterator):
+        if len(resources_by_sha1) == 1000:
+            # Send a request off for every 1000th sha1 collected
+            matched_count += matcher_func(
+                project=project,
+                resources_by_sha1=resources_by_sha1,
+                package_data_by_purldb_urls=package_data_by_purldb_urls,
+            )
+            resources_by_sha1 = defaultdict(list)
+            last_percent = pipes.log_progress(
+                logger,
+                resource_index,
+                resource_count,
+                last_percent,
+                increment_percent=10,
+                start_time=start_time,
+            )
+
+        resources_by_sha1[to_resource.sha1].append(to_resource)
+        if to_resource.path.endswith(".map"):
+            # Add sha1 of JS sources if we have a .map file
+            for js_sha1 in js.source_content_sha1_list(to_resource):
+                if len(resources_by_sha1) == 1000:
+                    matched_count += matcher_func(
+                        project=project,
+                        resources_by_sha1=resources_by_sha1,
+                        package_data_by_purldb_urls=package_data_by_purldb_urls,
+                    )
+                    resources_by_sha1 = defaultdict(list)
+                    last_percent = pipes.log_progress(
+                        logger,
+                        resource_index,
+                        resource_count,
+                        last_percent,
+                        increment_percent=10,
```

Comment:
This whole logic is too complex. Could it be simplify by chunking instead of depending on a harcoded `== 1000`.
See https://github.com/nexB/scancode.io/blob/main/scanpipe/pipes/vulnerablecode.py#L75 and https://github.com/nexB/scancode.io/blob/main/scanpipe/pipes/vulnerablecode.py#L205
for some examples of splitting the batch of item on several requests.
(We can move the `chunked` function out of the vulnerable code pipes module for the generic one if needed)

Reply:
@tdruez I tried using `chunked`, but it relies on list slicing to divide the iterable passed to it, which doesn't work on querysets. I've adapted this code from vulnerablecode (https://github.com/nexB/scancode.io/blob/781-purldb-matching-updates/scanpipe/models.py#L1612) and added it to the `CodebaseResource` queryset to return `CodebaseResource`s in groups of arbitrary sizes.